### DISCUSSION TITLE
Is xstate a good lib for dynamic forms

### DISCUSSION THREAD

**Discussion by johnsonthedev** [statelyai/xstate#1848] [Q&A]
Hi, I have the below case and was wondering if it is a good fit for xstate.

I have a form with 20-30 fields that have dependencies on each other. Here is an example:

I have three select fields: Country, State, City. State and City are disabled by default. If I select a country, the state field should become selectable—City after selecting a state.

Another example, I have two fields, a Radio Group, asking for contact options. If they say Yes, I want to show an input field for the email address.

```
const form = [
  {
    id: "contact",
    type: "radio-box",
    disabled: false,
    options: ["Yes", "No"],
  },
  {
    id: "email",
    type: "input-field",
    disabled: false,
    rules: [{ type: "visible", id: "contact", key: 0 }],
  },
]
```
Currently, I have an events object assigned to each form field and a function that I call with the onChange event.

But this becomes quite messy, as I have to run additional tasks like filtering the options, reset fields, validate user input, show/hide fields based on selection, and more.

I searched for a better solution to build such dependencies and wondered if xstate could help here.

Would you think it's a good idea to setup each field as a machine, or would that be too costly? If so, can I somehow connect each machine and manipulate its state?

Thx!

```
const form = [
  {
    id: "country",
    type: "select-box",
    disabled: false,
    options: [
      { value: "de", text: "Germany" },
      { value: "us", text: "USA" },
    ],
    events: {
      onChange: [
        { action: "enable-field", id: "state" },
        { action: "disable-field", id: "city" },
      ],
      isInvalid: [
          { action: "reset-field", id: "state" },
          { action: "reset-field", id: "city" },
      ]
    },
  },
  {
    id: "state",
    type: "select-box",
    disabled: true,
    options: [
      { value: "ba-wue", text: "Ba-Wü", parentId: "de" },
      { value: "ca", text: "Califoria", parentId: "us" },
    ],
    events: {
      onChange: [{ action: "enable-field", id: "city" }],
    },
  },
  {
    id: "city",
    type: "select-box",
    disabled: true,
    options: [
      { value: "hb", text: "Hamburg", parentId: "ba-wue" },
      { value: "sd", text: "San Diego", parentId: "ca" },
    ],
  },
]
```

**Comment by jaetask** [comment]
Hi @Johnson444 , I'm currently working on a composable xstate-form, visibility is a great usecase I hadn't considered. I'll add it to the roadmap now. Thanks

https://github.com/davidkpiano/xstate/discussions/1851

**Comment by ezhikov** [comment]
We recently migrated dynamic form from Redux to bunch of XState machines that were meant to control everything from large chunks of app to singular inputs. Overall experience was very delightful, but not without mistakes and caveats.

While XState gave us ability to swap and rewrite parts of application without touching others, we also got some data duplication and messed up communication between machines.

Our hierarchy is very straightforward and can be represented as such (from top to bottom):
- App - initial setup and data fetching. Also communication layer between first and second steps of our form. Steps pretty much independent and there is just data passing between first and second steps
- Step 1. Simple "choose one value from list or enter manually" with data fetching for next step.
- Step 2. This is most complex part. Depending on data fetched in first step in have many variations, so this part only gets data from first step and spawns number of machines to control fieldsets. Also it is responsible for sending form
- Fieldset is mostly data bus between fields and Step 2 machine. This is one of our mistakes, I'll elaborate a bit later.
- Field - pretty complex, but simplified (thanks xstate/inspect!) process that is reponsible for each field and it's multiple states (there is complex logic which involves juggling with two modes - edition and new value.

Basically, Step 2 have one or more Fieldsets and each Fieldet based on it's type can have five or more fields. Our first error was using internal communication of XState instead of React props and callbacks passing. In result we have to rely on internal xstate messages, that we have to type and explicitly accept in some cases. While it's not super bad, it would be better to organize communication via props and callbacks passed between react components (sometimes with context). This would make preserving state of inputs in different modes harder, but overall would allowed us to simplify communication, reduce data duplication and make some other readability optimizations. With this approach we also could've split some things out of xstate into simple components states (we could do this now, but it will be harder and less understandable).

On the bright side, treating each field with it's own machine and communicate between them via parents when needed gave us great power to "just react" to stimuli without actually concerning ourselves where does that stimuli comes from - is it user doing something, or it is internal matter or it is some chain reaction to something happened somewhere near.

In your example it could be represented as Form having many Fields. When something happens Field will send message to Form and Form will decide how to react - either send it to some other Fields, broadcast on every Field or do something else entierly.

**Reply by Andarist** [reply]
Comparing your current approach and the proposed one for a more complex task like the one you have described would make a great blog post! Just saying 😉

**Reply by ezhikov** [reply]
I'm already few month (actually spending few hours a week) into writing a piece about all that transition and self reflection, but I'm not particularly good with writing, especially that I decided to write it in English (not my native language), so I'll get there eventually, but not sure when.

**Reply by Andarist** [reply]
> I'm not particularly good with writing, especially that I decided to write it in English (not my native language)

Both are true for myself - but if u ever need help with reviewing ur article, let me know and ill do my best to help out

**Reply by jingoldby** [reply]
I would do terrible things for a repo of this setup

**Reply by sambokai** [reply]
@ezhikov A blog article would be greatly appreciated. Your english is absolutely fine.

**Reply by ezhikov** [reply]
So, it is still a very small draft and every time I sit to write anything meaningful out of it, I start to forget not only english words, but native language too. I still hope that I will be able to write something out of it.
